### PR TITLE
Adds application gateway

### PR DESCRIPTION
**Pull request by pabrus**
This PR is adding Azure Application Gateway to the reference hub-spoke implementation
Depends on https://github.com/mspnp/caf-secure-amp-infra/pull/15

### CODE REVIEW

File: deploy/caf-secure-ui.json [hunk 1]
```diff
@@ -384,10 +384,10 @@
                                 }
                             },
                             {
-                                "name": "ptSubnetAddressPrefixForAppGw",
+                                "name": "ptSubnetAddressPrefixForGw",
                                 "type": "Microsoft.Common.TextBox",
-                                "label": "Subnet prefix for Azure Application Gateway",
-                                "toolTip": "Subnet address prefix for Azure Application Gateway for the hub network",
+                                "label": "Subnet prefix for the gateway",
+                                "toolTip": "Subnet address prefix for the gateway for the hub network",
```
Comment (ckittel) [inline deploy/caf-secure-ui.json:390]: Not sure this text (or the previous text) makes sense in the context.  What is a "gateway _for the hub network_ mean?"  App Gateways are for spoke networks.

File: deploy/caf-secure-ui.json [hunk 2]
```diff
@@ -448,6 +448,46 @@
                                     "regex": "^([0-9]{1,3}\\.){3}[0-9]{1,3}(\/([0-9]|[1-2][0-9]|3[0-2]))?$",
                                     "validationMessage": "Please provide a valid IP address prefix"
                                 }
+                            },
+                            {
+                                "name": "ptSubnetAddressPrefixForAppGw",
+                                "type": "Microsoft.Common.TextBox",
+                                "label": "Subnet prefix for Azure Application Gateway",
+                                "toolTip": "Subnet address prefix for Azure Application Gateway for the hub network",
```
Comment (ckittel) [inline deploy/caf-secure-ui.json:456]: Not sure this text (or the previous text) makes sense in the context.  What is a "gateway _for the hub network_ mean?"  App Gateways are for spoke networks.

File: deploy/caf-secure-ui.json [hunk 3]
```diff
@@ -471,10 +511,10 @@
                                 }
                             },
                             {
-                                "name": "ptSubnetAddressPrefixForAppGw",
+                                "name": "ptSubnetAddressPrefixForGw",
                                 "type": "Microsoft.Common.TextBox",
-                                "label": "Subnet prefix for Azure Application Gateway",
-                                "toolTip": "Subnet address prefix for Azure Application Gateway for the spoke network",
+                                "label": "Subnet prefix for the gateway",
```
Comment (ckittel) [inline deploy/caf-secure-ui.json:516]: "gateway" might be best left as "Azure Application Gateway" to be clearer about what's being deployed.

File: deploy/auxiliary/caf-secure-networking-hub.json [hunk 4]
```diff
@@ -292,6 +320,153 @@
...
+                        {
+                            "type": "Microsoft.Network/applicationGateways",
+                            "apiVersion": "2020-05-01",
+                            "name": "[variables('appGwName')]",
+                            "location": "[resourceGroup().location]",
+                            "zones": "[if(not(empty(parameters('appGwZones'))), parameters('appGwZones'), json('null'))]",
+                            "dependsOn": [
+                                "[resourceId('Microsoft.Network/virtualNetworks', variables('hubNetworkName'))]",
+                                "[resourceId('Microsoft.Network/publicIPAddresses', variables('appGwIpName'))]"
+                            ],
```
Comment (ckittel) [inline deploy/auxiliary/caf-secure-networking-hub.json:349]: Really surprised to see this in the hub.  Is this guidance from someplace?  Typically an App Gateway is a workload level concern that exists in a spoke.

**Review by ckittel — CHANGES_REQUESTED** [review]
Before I get too far into the review, just want to validate a pretty significant decision to have the hub contain an App Gateway, vs the workload spokes.

**Comment by pabrus** [comment]
@ckittel I have done some changes now. Could you please check it again? Thanks

**Comment by ckittel** [comment]
Thanks @pabrus -- I'll add this to my review list for tomorrow.

**Comment by ckittel** [comment]
Just notes from our call (from what I can remember):

* Tooltip saying something like "Availability zones are only available in certain Azure regions." on the Azure region and the items where availability zones are selected.
* Availability zones (1,2,3) are selected by default.
* NSG on Azure Bastion -- no need for config option for this.
* NSG on App GW (80 from internet + required) -- no need for config option for this.
* OWASP 3.2 -- no need for config option for this.
* multiple IPs on FW (at least 2, 3 would be good as well) -- no need for config option for this.

**Comment by pabrus** [comment]
> * Availability zones (1,2,3) are selected by default.
This one can't be done 😞 defaultValue only allows one option even if the dropdown is set to be multiple.

File: README.md [hunk 5]
```diff
@@ -65,7 +65,7 @@ Azure Onboarding is implemented using a single deployment script that will creat
 
 Use the following link to deploy it to a greenfield subscription:
 
-[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fmspnp%2Fcaf-secure-amp-infra%2Fmain%2Fdeploy%2Fcaf-secure-deploy.json/createUIDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fmspnp%2Fcaf-secure-amp-infra%2Fmain%2Fdeploy%2Fcaf-secure-ui.json)
+[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fmspnp%2Fcaf-secure-amp-infra%2Fpabrus%2Fappgw%2Fdeploy%2Fcaf-secure-deploy.json/createUIDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fmspnp%2Fcaf-secure-amp-infra%2Fpabrus%2Fappgw%2Fdeploy%2Fcaf-secure-ui.json)
```
Comment (pabrus) [inline README.md:68]: Update branch before merging

File: deploy/caf-secure-deploy.json [hunk 6]
```diff
@@ -225,7 +263,7 @@
         }
     },
     "variables": {
-        "templateRootUri": "https://raw.githubusercontent.com/mspnp/caf-secure-amp-infra/main/deploy/auxiliary/",
+        "templateRootUri": "https://raw.githubusercontent.com/mspnp/caf-secure-amp-infra/pabrus/appgw/deploy/auxiliary/",
```
Comment (pabrus) [inline deploy/caf-secure-deploy.json:266]: Update branch before merge

**Review by ckittel — APPROVED** [review]
Besides your branch change in the URLs, good to go.  Thanks for making the changes.  Bummer about the multi-select defaults.

**Pull request merged**